### PR TITLE
LibJS/Bytecode: ElimiateLoads lost knowledge via ternaries

### PR DESCRIPTION
**Pull request by sec05**
When certain ternary statements are passed a block barrier can be created resulting in knowledge being lost between blocks. Further, this causes access to a not initialized register. This is fixed by changing `preform()` to walk the blocks in CFG order and replace their entries with new blocks after `eliminate_loads()` is preformed. Additionally, the register rerouting table must be made static in order for its order to be preserved between function calls.

### CODE REVIEW

File: Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp [hunk 1]
```diff
@@ -29,7 +29,7 @@ static NonnullOwnPtr<BasicBlock> eliminate_loads(BasicBlock const& block, size_t
 
     auto new_block = BasicBlock::create(block.name(), block.size());
     HashMap<size_t, Register> identifier_table {};
-    HashMap<u32, Register> register_rerouting_table {};
+    static HashMap<u32, Register> register_rerouting_table {};
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp:32]: This will need to be cleared when the executable changes, otherwise we get some really weird behavior

File: Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp [hunk 2]
```diff
@@ -182,15 +182,14 @@ void EliminateLoads::perform(PassPipelineExecutable& executable)
 {
     started();
 
-    // FIXME: If we walk the CFG instead of the block list, we might be able to
-    //        save some work between blocks
-    for (auto it = executable.executable.basic_blocks.begin(); it != executable.executable.basic_blocks.end(); ++it) {
-        auto const& old_block = *it;
+    auto cfg = executable.cfg.release_value();
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp:185]: Make sure we generate a CFG before just releasing it
For that look into the PassManager and where it is created and what other passes do

**Comment by Hendiadyoin1** [comment]
Also Just noticed, can you rephrase the commit message to be in an imperative style ideally describing the solution

File: Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp (hunk 1 above)
Comment (sec05) [reply Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp:32]: Cleared as in resetting the hash map? Like clear all the previous data in it?

Comment (Hendiadyoin1) [reply Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp:32]: Yes, as in resetting it, or calling `clear()` on it

File: Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp [hunk 3]
```diff
@@ -29,9 +29,10 @@ static NonnullOwnPtr<BasicBlock> eliminate_loads(BasicBlock const& block, size_t
 
     auto new_block = BasicBlock::create(block.name(), block.size());
     HashMap<size_t, Register> identifier_table {};
-    HashMap<u32, Register> register_rerouting_table {};
+    static HashMap<u32, Register> register_rerouting_table {};
 
     for (auto it = InstructionStreamIterator(block.instruction_stream()); !it.at_end();) {
+        register_rerouting_table.clear();
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp:35]: Now you are clearing it per instruction, so it will be always empty

**Comment by sec05** [comment]
I am not sure why the commit lint failed since I have not changed it at all from previous iterations. It says there's a http error in the commit linter job

File: Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp [hunk 4]
```diff
@@ -29,8 +29,8 @@ static NonnullOwnPtr<BasicBlock> eliminate_loads(BasicBlock const& block, size_t
 
     auto new_block = BasicBlock::create(block.name(), block.size());
     HashMap<size_t, Register> identifier_table {};
-    HashMap<u32, Register> register_rerouting_table {};
-
+    static HashMap<u32, Register> register_rerouting_table {};
+    register_rerouting_table.clear();
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp:33]: And now it's still cleared per block, so back on square one

**Review by Hendiadyoin1 — COMMENTED** [review]
You can also mark the older comments resolved

File: Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp (hunk 4 above)
Comment (sec05) [reply Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp:33]: Im just confused where it should be cleared? After its definition is all the instructions. Should it be cleared on the call to `perform()`? Some insight would be very much appreciated. Thank you!

Comment (Hendiadyoin1) [reply Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp:33]: Yes, it should be cleared on a call to `perform`, for that you will need to move the variable declaration out of the function

File: Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp [hunk 5]
```diff
@@ -181,16 +180,20 @@ static NonnullOwnPtr<BasicBlock> eliminate_loads(BasicBlock const& block, size_t
 void EliminateLoads::perform(PassPipelineExecutable& executable)
 {
     started();
+    static HashMap<u32, Register> register_rerouting_table {};
```
Comment (sec05) [inline Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp:183]: We shouldn't need to call `clear()` right? Since it's being defined to an empty map on every call to preform.

Comment (Hendiadyoin1) [reply Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp:183]: I'd moved it up out of both functions but this should be fine as well,
note that #20122 will make this effort for naught

File: Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp [hunk 6]
```diff
@@ -181,16 +180,20 @@ static NonnullOwnPtr<BasicBlock> eliminate_loads(BasicBlock const& block, size_t
...
+
+    auto pm = make<PassManager>();
+    pm->add<GenerateCFG>();
+    pm->perform(executable);
+    auto cfg = executable.cfg.release_value();
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp:188]: please move this into the main pass manager and not do it here ad-hoc

**Comment by MacDue** [comment]
Sadly this pass (and all other passes) were removed in #20122, so this PR won't be mergeable anymore. Sorry :(
